File: BackEnd/Domain/Aggregates/Department.cs
using Domain.Events.DepartmentCreated;
using SharedKernel.Aggregate;

namespace Domain.Aggregates;

public class Department : AggregateRoot<Guid>
{
    public string Name { get; set; } = string.Empty;
    public ICollection<User> Users { get; set; } = new List<User>();

    public static Department Create(string name)
    {
        return new Department
        {
            Id = Guid.CreateVersion7(),
            Name = name
        };
    }

    public void SetUsers(List<User> users)
    {
        Users = users;
    }

    public void DepartmentCreated()
    {
        AddDomainEvent(new DepartmentCreated
        {
            DepartmentId = Id
        });
    }

    public void Update(string name)
    {
        Name = name;
    }
}

File: flutter/lib/domain/aggregates/department/department.dart
import 'package:dart_mappable/dart_mappable.dart';
import 'package:enterprise_management/domain/aggregates/user/user.dart';
import 'package:enterprise_management/infrastructure/data/dtos/users/user_dto.dart';
import 'package:enterprise_management/shared_kernel/aggregate/aggregate_root.dart';

part 'department.mapper.dart';

@MappableClass()
class Department extends AggregateRoot<String> with DepartmentMappable {
  Department({
    required super.id,
    required this.name,
    this.isActive = false,
    required this.users,
  });

  final String name;
  final List<User> users;
  final bool isActive;
}
